File: java/src/main/java/io/dure/tree/RightSideView.java
package io.dure.tree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class RightSideView {
    public List<Integer> rightSideView(TreeNode root) {
        Queue<TreeNode> q = new LinkedList<>();
        List<Integer> views = new ArrayList<>();
        int current = 0;
        int next = 0;
        if (root != null) {
            q.offer(root);
            current = 1;
        }
        while (!q.isEmpty()) {
            TreeNode currNode = q.poll();
            current--;
            if (currNode.left != null) {
                q.offer(currNode.left);
                next++;
            }
            if (currNode.right != null) {
                q.offer(currNode.right);
                next++;
            }
            if (current == 0) {
                current = next;
                next = 0;
                views.add(currNode.val);
            }
        }
        return views;
    }
}

File: go/tree/rightSideView.go
package tree

func rightSideView(root *TreeNode) []int {
	var views []int
	var q []*TreeNode
	current := 0
	next := 0
	if root != nil {
		q = append(q, root)
		current = 1
	}
	for len(q) != 0 {
		currNode := q[0]
		q = q[1:]
		current--
		if currNode.Left != nil {
			q = append(q, currNode.Left)
			next++
		}
		if currNode.Right != nil {
			q = append(q, currNode.Right)
			next++
		}
		if current == 0 {
			current = next
			next = 0
			views = append(views, currNode.Val)
		}
	}
	return views
}
